File: app/src/main/java/com/left/gank/ui/discovered/video/VideoPresenter.java
package com.left.gank.ui.discovered.video;

import android.business.domain.Gank;
import android.content.Context;

import com.left.gank.bean.GankResult;
import com.left.gank.config.MeiziArrayList;
import com.left.gank.mvp.source.remote.GankDataSource;
import com.socks.library.KLog;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * Create by LingYan on 2017-01-03
 */

public class VideoPresenter extends VideoContract.Presenter {
    private GankDataSource mTask;
    private VideoContract.View mModelView;

    VideoPresenter(Context context, VideoContract.View view) {
        super(context, view);
    }

    private void fetchData(int page) {
        Observable<GankResult> observable;
        if (MeiziArrayList.getInstance().isOneItemsEmpty()) {
            observable = mTask.fetchVideoAndImages(page, 20);
        } else {
            observable = mTask.fetchVideo(page, 20);
        }

        observable.subscribe(new Observer<GankResult>() {
            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(GankResult gankResult) {
                List<Gank> list = gankResult.getResults(), mModelView;
            }

            @Override
            public void onError(Throwable e) {
                KLog.e(e);
            }

            @Override
            public void onComplete() {
            }
        });
    }

    @Override
    public void destroy() {

    }

    @Override
    protected void onDestroy() {

    }
}

File: app/src/main/java/com/left/gank/ui/discovered/video/VideoPresenter.kt
package com.left.gank.ui.discovered.video

import android.business.api.GankServer
import android.business.domain.PageEntity
import android.business.domain.Solid
import android.business.observer.ManagerObserver
import android.content.Context
import com.uber.autodispose.ObservableSubscribeProxy
import retrofit2.Response

/**
 * Create by LingYan on 2017-01-03
 */

class VideoPresenter internal constructor(context: Context, view: VideoContract.View) : VideoContract.Presenter(context, view) {

    override fun loadVideo(refresh: Boolean, useProgress: Boolean, page: Int) {
        if (isDestroy) {
            return
        }

        GankServer.with()
                .api()
                .video(refresh, page, INIT_LIMIT)
                .doOnSubscribe { showProgress(useProgress) }
                .doFinally { hideProgress() }
                .`as`<ObservableSubscribeProxy<Response<PageEntity<Solid>>>>(bindLifecycle<Response<PageEntity<Solid>>>())
                .subscribe(object : ManagerObserver<PageEntity<Solid>>(requestTag, obtainObserver()) {
                    override fun onSuccess(entity: PageEntity<Solid>?) {
                        if (isViewLife) {
                            if (entity != null) {
                                view.loadVideoSuccess(page, entity.results)
                                return
                            }
                            view.loadVideoFailure(page, errorTip)
                        }
                    }

                    override fun onFailure(e: Throwable) {
                        e.printStackTrace()
                        if (isViewLife) {
                            view.loadVideoFailure(page, errorTip)
                        }
                    }
                })
    }

    override fun onDestroy() {

    }

    companion object {
        private const val INIT_LIMIT = 20
    }
}
